### PR TITLE
Added method to load commands dynamically

### PR DESCRIPTION
**Pull request by bosunski**
Pull Request for Issue #1  .

### Summary of Changes
Added a method inside libraries/src/Application/ConsoleApplication.php to make loading of commands dynamic.

### Testing Instructions
Add a command class inside  libraries/src/Console/ and it should be loaded as a command when you run `php cli/joomla.php` 

### Expected result
Commands are loaded

### Actual result
Commands are Loaded

### CODE REVIEW

**Comment by mbabker** [comment]
👎 on this approach.  We should be favoring a configuration over convention solution, with that said here are a couple of my concerns:

1) This should be handled in the [dependency injection container's service definition](https://github.com/joomla-projects/gsoc18_cli_update/blob/69eaf4ed104b9dda69ff99dd99344ce507d9d274/libraries/src/Service/Provider/Application.php#L91-L112) versus in the application class, especially as we want to enable classes to have required dependencies, which leads to point 2...

2) This should really be calling `$container->buildSharedObject($classname);` versus just trying to direct instantiate an object.  As is you're breaking the dependency injection principal in place with the session garbage collection command (note how the command is [configured as a service](https://github.com/joomla-projects/gsoc18_cli_update/blob/master/libraries/src/Service/Provider/Console.php)), and the ability to have commands configured as services.  Getting the command through the dependency injection container's `buildSharedObject` lets the container handle automatically resolving constructor dependencies or use defined container services and eliminates the need for you to completely remove the constructor like you did.

**Comment by bosunski** [comment]
Thanks @mbabker , I will update!

**Comment by bosunski** [comment]
@mbabker if I get it correctly, what you're saying is that I shouldn't instantiate the class (which I did through ReflectionClass) like that, but rather I should pass it through the Container. Using `buildSharedObject` Method.

Am I correct?

You can please share more on this on how it can be achieved.

**Comment by alikon** [comment]
maybe i'm too much db's addicted... i'm telling wrong if we can be dynamic, on what we read on some kind of new table `cli_commands`

**Comment by alikon** [comment]
> .... we still need a good process for grabbing extension commands (which by definition have to be dynamic).

that was my original   question ,  how we can manage commands withouth hardcoding ?

**Comment by mbabker** [comment]
Right now for extensions they need a plugin in either the `system` or `console` plugin groups (the latter being explicit to the console app).  If you've got a better idea I'm all eyes, but the best option I could come up with was an early hook in the system for extensions to add their commands.

**Comment by alikon** [comment]
console plugin is a good compromise 4 me... expecially if we can have some kind of cron interface...

**Comment by mbabker** [comment]
It should go [here](https://github.com/joomla-projects/gsoc18_cli_update/blob/69eaf4ed104b9dda69ff99dd99344ce507d9d274/libraries/src/Service/Provider/Application.php#L91-L112).

**Comment by bosunski** [comment]
Getting this from the container after putting the code there:
`Failed to start application: Key Joomla\Console\Application is protected and can't be overwritten.

And also i don't get much of error reports for debugging, all I get is `Error` even if its a syntax error.

**Comment by bosunski** [comment]
@mbabker I have been able to update the `ServiceProvider` like you said, but the container still cannot resolve the Constructor dependency for `SessionGcCommand.` 
You should check latest commit to see what I did [here](https://github.com/joomla-projects/gsoc18_cli_update/pull/3/commits/b66a4bf1386c6bba09bb70a7192c42a1fd202d27)

File: libraries/src/Service/Provider/Application.php [hunk 1]
```diff
@@ -106,6 +106,22 @@ function (Container $container)
...
+
+					foreach ($paths as $key => $path) {
+						$namespaces[] = str_replace(
+							['.php', DIRECTORY_SEPARATOR],
+							['', '\\'],
+							explode($srcDir, $path)[1]
+						);
+					}
+					foreach ($namespaces as $key => $command) {
+							$className = '\Joomla\CMS' . $command;
```
Comment (mbabker) [inline libraries/src/Service/Provider/Application.php]: Remove the leading slash.  When PHP classes are used as the service IDs, the leading slash isn't present.

**Comment by mbabker** [comment]
Personally I'm still not convinced that the core commands need to be dynamically loaded in this way, IMO it's really error prone (especially if a class is added to that directory that's not a command class).

Actually, this can't work for the session command.  The session service relies on `Joomla\CMS\Factory::$application` being seeded, that doesn't happen until after the console application has been instantiated and this forces all core commands (and their dependencies) to be instantiated with the application class.  So in this case it needs to stay with the explicit configuration that is in place now versus trying to dynamically load all of the core commands and in essence breaking the ability to use dependency injection.

**Comment by alikon** [comment]
don't want to start a philosophical discussion on what should be core  commands  and what not,
so i want to stay pratical, session commands apart, do you think  we should consider a 2 way solution
lets's say core commands -> "static"  and non-core commands ->  "dynamic"  as in essence is proposed with this pr ?

**Comment by mbabker** [comment]
For me core commands are everything in the `Joomla\CMS` namespace.  Because we're shipping the `libraries/src` directory, we know what should be there and if the resources aren't there then there's a bigger problem.

Dynamic commands would be something like the [yet to be migrated Smart Search indexer script](https://github.com/joomla-projects/gsoc18_cli_update/blob/master/cli/finder_indexer.php).  Because that should belong inside `com_finder` as a Command class, not in the CLI directory as a standalone script, and as we well know components aren't loaded until they're explicitly told to be.

So long and short, if it's in the core libraries, we can safely rely on hardcoding the configuration references as is the case now.  It's anything that's based on extensions that we need an answer to the dynamic thing on, and like I pointed out before right now the solution is have a plugin register the commands at the console app's before execute event.

**Comment by alikon** [comment]
cannot say that you sold me this easily :smiley: 
... i can surely say a gain for @bosunski ... 
as at least he need to re-think ... a good taste of real life

**Comment by bosunski** [comment]
@alikon what are you suggesting?

**Comment by bosunski** [comment]
@mbabker how about we separate folders for core Commands and other commands?

**Comment by bosunski** [comment]
Or what other ways do you think we can move forward on this?
What are the steps you would recommend so we can move.

**Comment by mbabker** [comment]
They will be separate.  The core commands will all be in the `libraries/src/Console` directory.  Extension commands will be in their respective extension directories (i.e. when the indexer script I linked earlier is updated the command class should be in the `administrator/components/com_finder` directory.

The intent is not for extension developers to put their command classes in some arbitrary directory like we suggest in Joomla 3.x with putting scripts in the root `cli` directory.

**Comment by bosunski** [comment]
Okay. That intent seems clearer.
So, what should we be considering as my next of first step to that?

**Pull request closed without merging**